src/main/java/com/google/devtools/build/lib/remote/common/RemotePathResolver.java

```diff
@@ -0,0 +1,194 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote.common;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.devtools.build.lib.actions.ActionInput;
+import com.google.devtools.build.lib.actions.ActionInputHelper;
+import com.google.devtools.build.lib.exec.SpawnRunner.SpawnExecutionContext;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.IOException;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link RemotePathResolver} is used to resolve input/output paths for remote execution from
+ * Bazel's internal path, or vice versa.
+ */
+public interface RemotePathResolver {
+
+  /**
+   * @return the {@code workingDirectory} for a remote action.
+   */
+  @Nullable
+  String getWorkingDirectory();
+
+  /**
+   * @return a {@link SortedMap} which maps from input paths for remote action to {@link
+   * ActionInput}.
+   */
+  SortedMap<PathFragment, ActionInput> getInputMapping(SpawnExecutionContext context)
+      throws IOException;
+
+  /**
+   * Resolves the output path relative to input root for the given {@link Path}.
+   */
+  String resolveOutputPath(Path path);
```

Comment:
I was about to say that it's probably better to have only one overload for `resolveOutputPath()` and the one taking an execpath seems to be obviously the best, but then I looked at `RemoteCache.UploadManifest()`, which is AFAIU the only place where it's needed and realized that removing `Path` from that call would be quite a bit of work.

It seems always to come from `RemoteSpawnRunner.resolveActionInputs()`, so I think it's still a good idea to do because then you plumb the Path representing the execroot into `UploadManifest` only wrapped in `RemotePathResolver`, but it's probably not for this change. WDYT?

Reply:
You are right. I wanted to only have `resolveOutputPath(PathFragment execPath)` but then realized that `upload` methods take `Path` for output files and it need more work to replace them with `execPath`.

Let's refactor this along with changing `RemotePathResolve` to be a member of `RemoteCache` together in a separate PR!